File: src/_01_CreationalPattern/_01_02_FactoryMethod/Ship.js
export default class Ship {
  /** @type { string } */
  #name;

  /** @type { string } */
  #type;

  /** @type { string } */
  #color;

  /** @type { string } */
  #logo;

  /**
   * @param { string } name
   * @param { string } type
   * @param { string } color
   * @param { string } logo
   */
  constructor(name, type, color, logo) {
    this.#name = name;
    this.#type = type;
    this.#color = color;
    this.#logo = logo;
  }

  /** @returns { string } */
  get name() {
    return this.#name;
  }
  
  /** @param { string } name */
  set name(name) {
    this.#name = name;
  }

  /** @returns { string } */
  get type() {
    return this.#type;
  }

  /** @param { string } type */
  set type(type) {
    this.#type = type;
  }

  /** @returns { string } */
  get color() {
    return this.#color;
  }

  /** @param { string } color */
  set color(color) {
    this.#color = color;
  }

  /** @returns { string } */
  get logo() {
    return this.#logo;
  }

  /** @param { string } logo */
  set logo(logo) {
    this.#logo = logo;
  }
  
  print() {
    console.log(`
      Ship: {
        name: "${this.#name}",
        type: "${this.#type}",
        color: "${this.#color}",
        logo: "${this.#logo}",
      }
    `);
  }
}

File: src/_01_CreationalPattern/_01_03_AbstractFactory/java/before/Ship.java
package _01_CreationalPattern._01_03_AbstractFactory.java.before;

public class Ship {

  private String name;
  private String type;
  private String color;
  private String logo;
  private Anchor anchor;
  private Wheel wheel;

  public String getName() {
    return this.name;
  }
  public void setName(String name) {
    this.name = name;
  }

  public String getType() {
    return this.type;
  }
  public void setType(String type) {
    this.type = type;
  }

  public String getColor() {
    return this.color;
  }
  public void setColor(String color) {
    this.color = color;
  }

  public String getLogo() {
    return this.logo;
  }
  public void setLogo(String logo) {
    this.logo = logo;
  }

  public Anchor getAnchor() {
    return this.anchor;
  }
  public void setAnchor(Anchor anchor) {
    this.anchor = anchor;
  }

  public Wheel getWheel() {
    return this.wheel;
  }
  public void setWheel(Wheel wheel) {
    this.wheel = wheel;
  }
}
